### ISSUE TITLE
Uploaded Resource files randomly disappeared from ckan.

### ISSUE THREAD

**Issue by Ajaysamdyan** [ckan/ckan#5298]
**General information:**

Ckan version 2.7.2
Ckan-service-provider 0.0.6
Ckan Extensions: datapusher, recline_view, datastore, resource_proxy

**Issue description:**

 While continuously adding resource files to CKAN, added files disappeared or deleted without permission from ckan-dataset. Although no logs are produced, when the resource files are disappeared.
Possible reasons:
1.	Possibly its happened due to job failure of datapusher.
2.	Inside datapusher ext. sometimes the asynchronous job action is interrupted while adding another resource file due to which file gets disappeared from CKAN. I have studied code of “Datapusher” extension and found the suspected code at https://github.com/ckan/datapusher/blob/0.0.6/datapusher/jobs.py. Here, an asynchronous call of push_to_datastore() is performed. This function might create this problem.
This will be a big impact in CKAN functionality to upload resource file and it may impact other extensions as well.
I want to confirm if there is any way to define the synchronous job method in “job.py” to improvise this behavior ok CKAN functionality.
Is there any other possible reason for this issue.

**Comment by smotornyuk** [comment]
It's quite an old issue - datasets lost their resources during a concurrent update. If you are running datapusher service in multi-threads/processes, reducing the number of workers to just 1 may help. Otherwise, unfortunately, you have to implement some kind of mutex that prevents simultaneous updates of the same dataset.

Currently, there is an undergoing review of a new feature, that allows overcoming the mentioned issue - #4618  . But it's not yet merged into master and it won't be backported during upcoming patch-releases.

**Comment by jeanpommier** [comment]
Thanks @smotornyuk, very interesting information.
Just a question : why are you talking of concurrent update ? I run the sync in a regular basis, but my pb happens when nobody else is editing the organization, so this is not a concurrent update, right ?

**Comment by smotornyuk** [comment]
Maybe not quite a suitable term, but from those lines:
> sometimes the asynchronous job action is interrupted while adding another resource
I understood that you were editing the dataset(adding new resource) exactly at the moment when another resource was updating(ingesting into datastore, etc.). From the point of view of each resource - it's unrelated operation. But inside CKAN resource updated implemented as two-step operation `update_resource->update_parent_dataset`. So, from the point of view of parent package - it looks like two simultaneous updates, that's what I meant by `concurrent` update. 
And, as we are talking about it, when I saw similar issues before, next thing was causing it:

* first resource is updated(FINISHED) -> trigger update of parent datset(NOT FINISHED)
* first update fetches all the dataset details
* second resource updated(in parallel, using multiple workers from apache, uwsgi, gunicort, etc)(FINISHED) -> trigger update of parent dataset(NOT FINISHED)
* second update fetches all the dataset details
* **at this point we have two 'scheduled' updates of the same dataset. But the thing is, first update knows only about first resource, while second update knows about both resources**
* second update finished(why not - we cannot predict the order of thread execution) - there are two resources
* first update finished(update, that contains information only about first resource) - there is only one resource

**Comment by jeanpommier** [comment]
Oh, yeah, sorry, I thought it was an answer to #5293, I mixed things up. Do you think it might be related cause ?

**Comment by smotornyuk** [comment]
No, i almost sure those two issues are not related. Unfortunately I had no time yet to dig into image_url issue(
